pkg/operator/controllers/cloudproviderconfig/cloudproviderconfig_controller.go

```diff
@@ -0,0 +1,236 @@
+	return &CloudProviderConfigReconciler{
+		AROController: base.AROController{
+			Log:    Log,
+			Client: client,
+			Name:   ControllerName,
+		},
+	}
+}
+
+// Reconcile makes sure that the cloud-provider-config is healthy
+func (r *CloudProviderConfigReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	r.Log.Debug("reconcile ConfigMap openshift-config/cloud-provider-config")
+
+	instance, err := r.GetCluster(ctx)
+	if err != nil {
+		r.Log.Info(err)
+		return reconcile.Result{}, err
+	}
+
+	if !instance.Spec.OperatorFlags.GetSimpleBoolean(controllerEnabled) {
+		r.Log.Info("controller is disabled")
+		return reconcile.Result{}, nil
+	}
+
+	r.Log.Debug("running")
+	return reconcile.Result{}, r.updateCloudProviderConfig(ctx)
+}
+
+// SetupWithManager setup our manager
+func (r *CloudProviderConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.Log.Info("starting cloud-provider-config controller")
+
+	aroClusterPredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.GetName() == arov1alpha1.SingletonClusterName
+	})
+
+	cloudProviderConfigPredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.GetName() == cloudProviderConfigName.Name && o.GetNamespace() == cloudProviderConfigName.Namespace
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&arov1alpha1.Cluster{}, builder.WithPredicates(aroClusterPredicate)).
+		Watches(
+			&source.Kind{Type: &corev1.ConfigMap{}},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(cloudProviderConfigPredicate),
+		).
+		Named(ControllerName).
+		Complete(r)
+}
+
+func (r *CloudProviderConfigReconciler) updateCloudProviderConfig(ctx context.Context) error {
+	r.Log.Debug("checking openshift-config/cloud-provider-config")
+
+	cm := &corev1.ConfigMap{}
+	err := r.Client.Get(ctx, cloudProviderConfigName, cm)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			r.Log.Debug("the ConfigMap cloud-provider-config was not found in the openshift-config namespace")
+		}
+		return err
+	}
+
+	jsonConfig, ok := cm.Data["config"]
+	if !ok {
+		return fmt.Errorf("field config in ConfigMap openshift-config/cloud-provider-config is missing")
+	}
+
+	changed := false
+	var cpc azCloudProviderConfig
+	err = json.Unmarshal([]byte(jsonConfig), &cpc)
+	if err != nil {
+		return err
+	}
+
+	if cpc.DisableOutboundSNAT != nil && !*cpc.DisableOutboundSNAT {
+		r.Log.Info("updating openshift-config/cloud-provider-config disableOutboundSNAT from false to true")
+		*cpc.DisableOutboundSNAT = true
+		changed = true
+	} else if cpc.DisableOutboundSNAT == nil {
+		r.Log.Info("updating openshift-config/cloud-provider-config disableOutboundSNAT from nil to true")
+		truePointer := true
+		cpc.DisableOutboundSNAT = &truePointer
+		changed = true
+	} else {
+		r.Log.Debug("openshift-config/cloud-provider-config disableOutboundSNAT is set to true no changes needed")
```

Comment:
Could we just return after this debug statement? This way we don't have to keep the `changed` variable.

Reply:
Removing the changed variable would make this more wordy. We'd be marshaling JSON and catching errors in every branch.

Reply:
I think returning nil here and removing changed simplies the code. There would be three paths, two of which need to continue on (changed = true) and the third which can exit early. Since the two non terminal paths also have changed = true, there are really only two paths: we need to update or we need to return. The update code can be left outside of the conditional block and we always just return `r.Client.Update()` as the last line of this method. IMO we should try to avoid using tracking booleans such as changed as it adds mental load when reading the code. I tried this change locally and the tests still passed :)

That is of course, unless we need to also check if we can marshal the unchanged `cpc`, which I think we can omit if everything is fine (changed = false).

Reply:
I do not agree that having a changed boolean adds mental load to the code. IMO it makes this more readable.